v2/comments.go

```diff
@@ -69,15 +72,219 @@ func ExtractCommentTags(marker string, lines []string) map[string][]string {
+		parts := strings.SplitN(in, "//", 2)
+		return strings.TrimSpace(parts[0])
+	}
+
+	out := map[string][]Tag{}
+	for _, line := range lines {
+		line = strings.TrimSpace(line)
+		if len(line) == 0 {
+			continue
+		}
+		if !strings.HasPrefix(line, marker) {
+			continue
+		}
+		line = stripTrailingComment(line)
+		kv := strings.SplitN(line[len(marker):], "=", 2)
+		key := kv[0]
+		val := ""
+		if len(kv) == 2 {
+			val = kv[1]
+		}
+
+		tag := Tag{}
+		if name, args, err := parseTagKey(key, tagNames); err != nil {
+			return nil, err
+		} else if name != "" {
+			tag.Name, tag.Args = name, args
+			tag.Value = val
+			out[tag.Name] = append(out[tag.Name], tag)
+		}
+	}
+	return out, nil
+}
+
+// Tag represents a single comment tag.
+type Tag struct {
+	// Name is the name of the tag with no arguments.
+	Name string
+	// Args is a list of optional arguments to the tag.
+	Args []string
+	// Value is the value of the tag.
+	Value string
+}
+
+func (t Tag) String() string {
+	buf := bytes.Buffer{}
+	buf.WriteString(t.Name)
+	if len(t.Args) > 0 {
+		buf.WriteString("(")
+		for i, a := range t.Args {
+			if i > 0 {
+				buf.WriteString(", ")
+			}
+			buf.WriteString(a)
+		}
+		buf.WriteString(")")
+	}
+	return buf.String()
+}
+
+// parseTagKey parses the key part of an extended comment tag, including
+// optional arguments. The input is assumed to be the entire text of the
+// original input after the marker, up to the '=' or end-of-line.
+//
+// The tags argument is an optional list of tag names to match. If it is nil or
+// empty, all tags match.
+//
+// At the moment, arguments are very strictly formatted (see parseTagArgs) and
+// whitespace is not allowed.
+//
+// This function returns the key name and arguments, unless tagNames was
+// specified and the input did not match, in which case it returns "".
+func parseTagKey(input string, tagNames []string) (string, []string, error) {
+	parts := strings.SplitN(input, "(", 2)
+	key := parts[0]
+
+	if len(tagNames) > 0 {
+		found := false
+		for _, tn := range tagNames {
+			if key == tn {
+				found = true
+				break
+			}
+		}
+		if !found {
+			return "", nil, nil
+		}
+	}
+
+	var args []string
+	if len(parts) == 2 {
+		if ret, err := parseTagArgs(parts[1]); err != nil {
+			return key, nil, fmt.Errorf("failed to parse tag args: %v", err)
+		} else {
+			args = ret
+		}
+	}
+	return key, args, nil
+}
+
+// parseTagArgs parses the arguments part of an extended comment tag. The input
+// is assumed to be the entire text of the original input after the opening
+// '(', including the trailing ')'.
+//
+// At the moment this assumes that the entire string between the opening '('
+// and the trailing ')' is a single Go-style identifier token, but in the
+// future could be extended to have multiple arguments with actual syntax.  The
+// single token may consist only of letters and digits.  Whitespace is not
+// allowed.
+func parseTagArgs(input string) ([]string, error) {
```

Comment:
consider `+something("argument, including commas")` or `+something({"json": 1, "parseThis": { "ohno": true }})`

I dont think we have need for those, but I don't KNOW.  I also don't know if we have need for `+something(multiple, simple, args)` - so rather than make any decision now, I think we can just punt that until we hve a concrete use case.